### ISSUE TITLE
Crashed: com.TUSKit.Scheduler

### ISSUE THREAD

**Issue by m3t8d1tr** [tus/TUSKit#146]
Hi there.
After migration to version 3, we started getting crashes on devices with iOS 12.

```
Crashed: com.TUSKit.Scheduler
0  libobjc.A.dylib                0x40e0 object_isClass + 16
1  Foundation                     0xfb280 KVO_IS_RETAINING_ALL_OBSERVERS_OF_THIS_OBJECT_IF_IT_CRASHES_AN_OBSERVER_WAS_OVERRELEASED_OR_SMASHED + 64
2  Foundation                     0xf8b20 NSKeyValueWillChangeWithPerThreadPendingNotifications.llvm.5898781120377872701 + 304
3  Foundation                     0x726f4 -[NSProgress _setValueForKeys:settingBlock:] + 264
4  Foundation                     0x75ac4 -[NSProgress setTotalUnitCount:] + 128
5  Foundation                     0x818a4 +[NSProgress progressWithTotalUnitCount:] + 100
6  CFNetwork                      0x26c3c8 -[__NSCFURLSessionTask progress] + 120
7  TUSKit                         0x188a8 UploadDataTask.run(completed:) + 159 (UploadDataTask.swift:159)
8  TUSKit                         0x9184 closure #1 in Scheduler.checkProcessNextTask() + 372 (<compiler-generated>:372)
9  TUSKit                         0x1a160 thunk for @escaping @callee_guaranteed () -> () + 28 (<compiler-generated>:28)
10 libdispatch.dylib              0x5fa38 _dispatch_call_block_and_release + 24
```
Any ideas?

**Comment by rick51231** [comment]
As I see, we have the same issue.
Xcode (14.1) with 6s (12.4) simulator crashes with EXC_BAD_ACCESS on UploadDataTask.swift:159  :
`        observation = task.progress.observe(\.fractionCompleted) { [weak self] progress, _ in`

Didn't found any solution for now.

**Comment by m3t8d1tr** [comment]
Fix: 
<img width="996" alt="image" src="https://user-images.githubusercontent.com/11466493/204792455-f7fea064-2c43-4a8a-a765-48311eeda3fd.png">

**Comment by Acconut** [comment]
@m3t8d1tr Would you be able to open a PR with this patch?

**Comment by rick51231** [comment]
@m3t8d1tr solution is not fixed issue.
I figured out that this crash caused by late "observation" invalidation.
@Acconut Opened a PR for this https://github.com/tus/TUSKit/pull/150

**Comment by Acconut** [comment]
Thank you for the PR, @rick51231. We will have a look at it soon.

**Comment by donnywals** [comment]
I unfortunately do not have an iOS 12 device to verify this fix with but since the change looks pretty minor and no other things have broken as far as I’m aware, I have merged your pr @rick51231 we’ll make a 3.1.5 release that includes this fix soon. (in the next few days)
